zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineCache.kt

```diff
@@ -0,0 +1,144 @@
+package app.cash.zipline.loader
+
+import kotlinx.datetime.Clock
+import okio.ByteString
+import okio.FileSystem
+import okio.Path
+import okio.Path.Companion.toPath
+
+/**
+ * Zipline Cache State Machine
+ *
+ *            download
+ *              start
+ *         --------------->
+ *   ABSENT                DIRTY
+ *         <---------------
+ *    ^       download       | download
+ *    |        failed        | success
+ *    |                      |
+ *    |                      v
+ *    '------------------ READY
+ *           pruned
+ *
+ *
+ * ABSENT entries are not present in sqlite.
+ *
+ * This cache is safe for concurrent use. If multiple callers attempt to mark the same ABSENT file
+ * as DIRTY simultaneously, all but one will fail and return false.
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ */
+class ZiplineCache(
+  private val fileSystem: FileSystem,
+  private val directory: Path,
+  private val database: Database,
+  private val maxSizeInBytes: Int = 100 * 1024 * 1024, // 100mb
+) {
+  /**
+   * Returns true if the file was absent and is now DIRTY. The caller is now the owner of this file
+   * and may proceed to write the file to the file system.
+   */
+  fun setDirty(
+    sha256: ByteString,
+  ): Boolean = try {
+    database.cacheQueries.insert(
+      sha256_hex = sha256.hex(),
+      file_state = FileState.DIRTY,
+      size_bytes = 0L,
+      last_used_at_epoch_ms = Clock.System.now().toEpochMilliseconds()
```

Comment:
kotlinx.datetime is a large and unstable dependency to take on here solely for getting the current time. I was first going to comment that `Clock` should be a constructor parameter to allow for testing, but I'm tempted to just take a lambda of `epochMillis: () -> Int` rather than take on the whole dependency. Not quite sure yet.